File: python/epi/number_ways.py
#!/usr/bin/env python3
#
# Description:
#
# Given an NxM 2D grid, count the number of unique paths from the top-left
# corner to the bottom-right corner. Going up, down, left, and right are the
# only valid movements.

from rlib.testing_utils import expect

def number_of_ways(n, m):
    memo = [[0 for j in range(m)] for i in range(n)]
    for j in range(m): memo[0][j] = 1
    for i in range(n): memo[i][0] = 1
    for i in range(1, n):
        for j in range(1, m):
            memo[i][j] = memo[i - 1][j] + memo[i][j - 1]
    return memo[n - 1][m - 1]

def small_test():
    expect(number_of_ways(1, 1)).to_be(1)
    expect(number_of_ways(1, 5)).to_be(1)
    expect(number_of_ways(5, 1)).to_be(1)
    expect(number_of_ways(5, 5)).to_be(70)

if __name__ == '__main__':
    small_test()

File: EpiJv/src/main/java/moe/ijnji/epi/NumberWays.java
// Description:
//
// Given an R-C 2D grid, count the number of paths from (0, 0), the top-left
// corner, to (R, C), the bottom-right corner, and the only two movement options
// are moving right and moving down.

package moe.ijnji.epi;



public class NumberWays {

  public static int numberOfWays(int r, int c) {
    if (r == 0 || c == 0) return 1;
    int[][] memo = new int[r][c];
    for (int i = 0; i < r; i++) memo[i][0] = 1;
    for (int j = 0; j < c; j++) memo[0][j] = 1;
    for (int i = 1; i < r; i++) {
      for (int j = 1; j < c; j++) {
        memo[i][j] = memo[i - 1][j] + memo[i][j - 1];
      }
    }
    return memo[r - 1][c - 1];
  }

  private static void smallTest() {
    //expect(numberOfWays(0, 0)).toBe(1);
    //expect(numberOfWays(1, 1)).toBe(1);
    //expect(numberOfWays(2, 2)).toBe(2);
    //expect(numberOfWays(1, 5)).toBe(1);
    //expect(numberOfWays(5, 1)).toBe(1);
    //expect(numberOfWays(4, 3)).toBe(10);
    //expect(numberOfWays(5, 5)).toBe(70);
  }

  public static void main(String[] args) {
    smallTest();
  }

}